### PR TITLE
[CELEBORN-1567] Support throw FetchFailedException when Data corruption detected

### PR DESCRIPTION
**Pull request by cxzl25**
### Why are the changes needed?
https://github.com/apache/celeborn/pull/2655#pullrequestreview-2213124224

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
GA

### CODE REVIEW

File: client/src/main/java/org/apache/celeborn/client/read/CelebornInputStream.java [hunk 1]
```diff
@@ -665,38 +667,38 @@ private boolean fillBuffer() throws IOException {
         }
 
         return hasData;
-      } catch (IOException e) {
+      } catch (Exception e) {
```
Comment (SteNicholas) [inline client/src/main/java/org/apache/celeborn/client/read/CelebornInputStream.java:670]: Why not directly catch `LZ4Exception|ZstdException|IOException`

**Review by SteNicholas — APPROVED** [review]

**Review by mridulm — COMMENTED** [review]
Looks good to me.

Btw, support for checksum/validation of data would a good feature to add IMO ... there were corner cases where this helped catch issues in spark (instead of relying on compression/deserialization failing ... which need not always happen).

**Review by RexXiong — APPROVED** [review]
LGTM

**Comment by cfmcgrady** [comment]
I have a question: should we retry fetching another replication before throwing a FetchFailedException when the conf `celeborn.client.push.replicate.enabled` is set to true?

**Comment by cxzl25** [comment]
> I have a question: should we retry fetching another replication before throwing a FetchFailedException when the conf `celeborn.client.push.replicate.enabled` is set to true?

This is not necessarily safe, because the Task may have read part of the data, so it is safer to retry the Task. This is how Spark handles it.

**Comment by cxzl25** [comment]
> support for checksum/validation of data would a good feature

It looks like we've already done this.

https://github.com/apache/celeborn/blob/0ee3c3a4bdef3b97fca6be04f03ac36c2e12e1a6/client/src/main/java/org/apache/celeborn/client/compress/ZstdDecompressor.java#L69-L72

https://github.com/apache/celeborn/blob/0ee3c3a4bdef3b97fca6be04f03ac36c2e12e1a6/client/src/main/java/org/apache/celeborn/client/compress/Lz4Decompressor.java#L82-L85

**Review by FMX — APPROVED** [review]
LGTM.

**Review by mridulm — APPROVED** [review]

**Review by cfmcgrady — APPROVED** [review]
LGTM

**Comment by cfmcgrady** [comment]
Thank you, merging to main(v0.6.0)/branch-0.5(v0.5.2)/branch-0.4(v0.4.3).

**Pull request closed without merging**